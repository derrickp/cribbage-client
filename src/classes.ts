

export enum Common {
    ERROR = "error"
}