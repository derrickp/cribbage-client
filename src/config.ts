export const BASE_API_URL = "http://localhost:4567";

export const MAX_HAND_SIZE = 4;