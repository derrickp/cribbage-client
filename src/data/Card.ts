
export interface Card {
    name: string;
    suit: string;
    fullName: string;
    image: string;
    key: string;
}