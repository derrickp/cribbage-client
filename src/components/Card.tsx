
import * as React from "react";

import * as classNames from "classnames";
import { Draggable } from 'react-beautiful-dnd';

import { Card as DataCard } from "../data/Card";

export interface CardProps {
    card: DataCard;
}

export class Card extends React.Component<CardProps, any> {

    constructor(props: CardProps) {
        super(props);

        this.getCardClass = this.getCardClass.bind(this);
    }

    getCardClass(dragging: boolean): string {
        const drag = ({
            [CardClasses.DRAGGING]: dragging
        } as any);
        return classNames(CardClasses.CARD, drag);
    }

    render() {
        const card = this.props.card;
        return <Draggable key={card.fullName} draggableId={card.fullName}>
            {(provided: any, snapshot) => (
                <div>
                    <div
                        ref={provided.innerRef}
                        style={{...provided.draggableStyle}}
                        className={this.getCardClass(snapshot.isDragging)}
                        {...provided.dragHandleProps}
                    >
                        {card.fullName}
                    </div>
                    {provided.placeholder}
                </div>
            )}
        </Draggable>
    }
}

export enum CardClasses {
    CARD = "card",
    DRAGGING = "card-dragging"
}