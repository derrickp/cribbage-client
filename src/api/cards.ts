
import { BASE_CRIB_URL } from "../config";
import { Card } from "../data/Card";

export async function getCards(): Promise<Card[]> {
    const response = await fetch(BASE_CRIB_URL + "/cards");
    const cards = await response.json();
    return cards;
}