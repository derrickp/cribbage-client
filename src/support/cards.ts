
import { BASE_API_URL } from "../config";

import { Card } from "../data/Card";

export async function getCards(): Promise<Card[]> {
    const response = await fetch(BASE_API_URL + "/cards");
    const cards = await response.json();
    return cards;
}