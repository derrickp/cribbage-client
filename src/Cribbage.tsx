
import * as React from "react";

import * as classNames from "classnames";
import { DragDropContext, DropResult, DraggableLocation } from 'react-beautiful-dnd';

import { Card, sortByName, sortBySuit, sortOptions } from "./data/Card";
import { Option } from "./data/Option";
import { CardBucket } from "./components/CardBucket";
import { Common } from "./classes";

import { calculateScore } from "./api/score";
import { getCards } from "./api/cards";

export interface CribbageProps { }

export interface CribbageState {
    deck: Card[];
    sortOption: Option;
    hand: Card[];
    cut: Card[];
    isCrib: boolean;
    score: number;
}

export class Cribbage extends React.Component<CribbageProps, CribbageState> {
    constructor(props: CribbageProps) {
        super(props);
        const nameSortOption = sortOptions.find(option => option.value === "name");
        this.state = {
            deck: [],
            hand: [],
            cut: [],
            isCrib: false,
            score: 0,
            sortOption: nameSortOption
        };

        this.onDragEnd = this.onDragEnd.bind(this);
        this.calculateScore = this.calculateScore.bind(this);
        this.toggleIsCrib = this.toggleIsCrib.bind(this);
        this.sortCards = this.sortCards.bind(this);
    }

    async componentWillMount() {
        const deck = (await getCards()).sort((a, b) => sortByName(a, b));

        if (deck && deck.length) {
            this.setState({
                deck
            });
        }
    }

    async calculateScore() {
        const score = await calculateScore(this.state.hand, this.state.cut[0], this.state.isCrib);
        this.setState({
            score
        });
    }

    onDragEnd(result: DropResult) {
        // dropped outside the list
        if (!result.destination) {
            return;
        }

        const { deck, hand, cut } = moveCards(
            this.state.deck,
            this.state.hand,
            this.state.cut,
            result.source,
            result.destination
        );

        const sortFunc = this.state.sortOption.value === "name" ? sortByName : sortBySuit;
        const sortedDeck = deck.sort(sortFunc);

        this.setState({
            cut,
            hand,
            deck: sortedDeck,
        });
    }

    toggleIsCrib(event: React.ChangeEvent<HTMLInputElement>) {
        const isCrib = !this.state.isCrib;
        this.setState({
            isCrib
        });
    }

    sortCards(sortOption: Option, id: string) {
        let sortFunc: (a: Card, b: Card) => number = sortOption.value === "name" ? sortByName : sortBySuit;
        const deck = Array.from(this.state.deck).sort(sortFunc)
        this.setState({
            deck,
            sortOption
        });
    }

    render() {
        if (!this.state.deck || this.state.deck.length <= 0) {
            return <div>Loading deck...</div>;
        }

        const className = classNames(Classes.APP);
        const handError = this.state.hand.length > 4 ? Common.ERROR : "";
        const cutError = this.state.cut.length > 1 ? Common.ERROR : "";
        const calcDisabled = handError !== "" || cutError !== "" || this.state.hand.length === 0;
        return (
            <DragDropContext onDragEnd={this.onDragEnd}>
                <div className={className}>
                    <CardBucket key={DroppableIds.DECK} id={DroppableIds.DECK} name="Deck (available cards)" cards={this.state.deck} sortCards={this.sortCards} sortEnabled={true} sortOptions={sortOptions} sortValue={this.state.sortOption}>
                    </CardBucket>
                    <CardBucket key={DroppableIds.HAND} id={DroppableIds.HAND} name="Hand (max: 4)" cards={this.state.hand} additionalClass={handError} sortCards={this.sortCards} sortEnabled={false}>
                    </CardBucket>
                    <CardBucket key={DroppableIds.CUT} id={DroppableIds.CUT} name="Cut Card (max: 1)" cards={this.state.cut} additionalClass={cutError} sortEnabled={false}>
                    </CardBucket>
                    <div className={Classes.SCORE_CONTAINER}>
                        <label><input type="checkbox" checked={this.state.isCrib} onChange={this.toggleIsCrib}></input>Crib?</label>
                        <button disabled={calcDisabled} className={Classes.SCORE_BUTTON} onClick={this.calculateScore}>Get Score</button>
                        {<div className={Classes.SCORE}>{`Your current score is ${this.state.score}`}</div>}
                    </div>
                </div>
            </DragDropContext>
        );
    }
}

enum DroppableIds {
    DECK = "deck",
    HAND = "hand",
    CUT = "cut"
};

export enum Classes {
    APP = "app",
    SCORE_BUTTON = "score-button",
    SCORE = "score",
    SCORE_CONTAINER = "score-container"
};

// a little function to help us with reordering the result
function moveCards(initialDeck: Card[], initialHand: Card[], initialCut: Card[], sourceLocation: DraggableLocation, destinationLocation: DraggableLocation) {
    // Make a copy of our arrays
    const deck = Array.from(initialDeck);
    const hand = Array.from(initialHand);
    const cut = Array.from(initialCut);

    let sourceIndex: number = sourceLocation.index;
    let destinationIndex: number = destinationLocation.index;

    let moved: Card;

    // Find the card that has been moved.
    // Splice it out of the list it was in.
    switch (sourceLocation.droppableId) {
        case DroppableIds.DECK:
            moved = deck.splice(sourceIndex, 1)[0];
            break;
        case DroppableIds.HAND:
            moved = hand.splice(sourceIndex, 1)[0];
            break;
        case DroppableIds.CUT:
            moved = cut.splice(sourceIndex, 1)[0];
            break;
    }

    // Add it to the list it's going in
    switch (destinationLocation.droppableId) {
        case DroppableIds.DECK:
            deck.splice(destinationIndex, 0, moved);
            break;
        case DroppableIds.HAND:
            hand.splice(destinationIndex, 0, moved);
            break;
        case DroppableIds.CUT:
            cut.splice(destinationIndex, 0, moved);
            break;
    }

    return {
        deck,
        hand,
        cut
    }
}