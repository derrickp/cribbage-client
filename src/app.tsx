
import "whatwg-fetch";
import "core-js";

import * as React from "react";
import * as ReactDOM from "react-dom";

import { Cribbage } from "./Cribbage";

const element = document.getElementById("cribbage");

ReactDOM.render(<Cribbage />, element);