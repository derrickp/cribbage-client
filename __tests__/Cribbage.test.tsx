
import * as React from "react";

import { create } from "react-test-renderer";

import { DraggableLocation } from "react-beautiful-dnd";

import { Card } from "../src/data/Card";
import { Cribbage, DroppableIds, moveCards } from "../src/Cribbage";

import { createCards } from "./support/cards";

describe("Cribbage", () => {
    it("moves cards between deck and hand as expected", () => {
        const initialDeck = createCards("AS 2H 3H");
        const initialHand = createCards("5H");
        const initialCut = createCards("2S");

        const expectedDeck = [ initialDeck[1], initialDeck[2] ];
        const expectedHand = [ initialHand[0], initialDeck[0] ];

        const source: DraggableLocation = {
            droppableId: DroppableIds.DECK,
            index: 0
        };
        
        const destination: DraggableLocation = {
            droppableId: DroppableIds.HAND,
            index: 1
        };

        const { deck, hand, cut } = moveCards(initialDeck, initialHand, initialCut, source, destination);
        expect(initialDeck).toHaveLength(3); // Does not modify initial array
        expect(deck).toHaveLength(2); // Resulting deck has card removed
        expect(deck).toEqual(expectedDeck);
        expect(deck).not.toEqual(initialDeck);
    });

    it("renders the app", () => {
        const deck = createCards("AS 2H 3H");
        const component = create(<Cribbage deck={deck} />).toJSON();
        expect(component).not.toBe(undefined);
        expect(component).toMatchSnapshot();
    });
});