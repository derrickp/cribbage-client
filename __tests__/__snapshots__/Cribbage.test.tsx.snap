// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cribbage renders the app 1`] = `
<div
  className="app"
>
  <div
    className="list-container"
  >
    <div
      className="header"
    >
      <h4>
        Deck (available cards)
      </h4>
      <div
        className="Select Select--single is-searchable has-value"
        style={undefined}
      >
        <div
          className="Select-control"
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          style={undefined}
        >
          <span
            className="Select-multi-value-wrapper"
            id="react-select-2--value"
          >
            <div
              className="Select-value"
              style={undefined}
              title={undefined}
            >
              <span
                aria-selected="true"
                className="Select-value-label"
                id="react-select-2--value-item"
                role="option"
              >
                Name
              </span>
            </div>
            <div
              className="Select-input"
              style={
                Object {
                  "display": "inline-block",
                }
              }
            >
              <style
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "input#undefined::-ms-clear {display: none;}",
                  }
                }
              />
              <input
                aria-activedescendant="react-select-2--value"
                aria-describedby={undefined}
                aria-expanded="false"
                aria-haspopup="false"
                aria-label={undefined}
                aria-labelledby={undefined}
                aria-owns=""
                className={undefined}
                id={undefined}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={false}
                role="combobox"
                style={
                  Object {
                    "boxSizing": "content-box",
                    "width": "5px",
                  }
                }
                tabIndex={undefined}
                value=""
              />
              <div
                style={
                  Object {
                    "height": 0,
                    "left": 0,
                    "overflow": "scroll",
                    "position": "absolute",
                    "top": 0,
                    "visibility": "hidden",
                    "whiteSpace": "pre",
                  }
                }
              >
                
              </div>
            </div>
          </span>
          <span
            className="Select-arrow-zone"
            onMouseDown={[Function]}
          >
            <span
              className="Select-arrow"
              onMouseDown={[Function]}
            />
          </span>
        </div>
      </div>
    </div>
    <div
      className="card-list"
    >
      <div>
        <div
          aria-grabbed={false}
          className="card"
          draggable={false}
          onClick={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          style={
            Object {
              "pointerEvents": "auto",
              "transform": null,
              "transition": null,
            }
          }
          tabIndex={0}
        >
          AS
        </div>
      </div>
      <div>
        <div
          aria-grabbed={false}
          className="card"
          draggable={false}
          onClick={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          style={
            Object {
              "pointerEvents": "auto",
              "transform": null,
              "transition": null,
            }
          }
          tabIndex={0}
        >
          2H
        </div>
      </div>
      <div>
        <div
          aria-grabbed={false}
          className="card"
          draggable={false}
          onClick={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          style={
            Object {
              "pointerEvents": "auto",
              "transform": null,
              "transition": null,
            }
          }
          tabIndex={0}
        >
          3H
        </div>
      </div>
    </div>
  </div>
  <div
    className="list-container"
  >
    <div
      className="header"
    >
      <h4>
        Hand (max: 4)
      </h4>
    </div>
    <div
      className="card-list"
    />
  </div>
  <div
    className="list-container"
  >
    <div
      className="header"
    >
      <h4>
        Cut Card (max: 1)
      </h4>
    </div>
    <div
      className="card-list"
    />
  </div>
  <div
    className="score-container"
  >
    <label>
      <input
        checked={false}
        onChange={[Function]}
        type="checkbox"
      />
      Crib?
    </label>
    <button
      className="score-button"
      disabled={true}
      onClick={[Function]}
    >
      Get Score
    </button>
    <div
      className="score"
    >
      Your current score is 0
    </div>
  </div>
</div>
`;
